File: swift/all-your-base/Sources/AllYourBase/AllYourBase.swift
fileprivate func pow(base: Int, exponent: Int) -> Int {
  if exponent == 0 {
    return 1
  }

  var n = 1
  for _ in 1...exponent {
    n *= base
  }
  return n
}

public enum BaseError: Error {
  case invalidInputBase
  case invalidOutputBase
  case invalidPositiveDigit
  case negativeDigit
}

struct Base {
  private static func
  digits(baseIn: Int, digits: [Int], baseOut: Int) -> [Int] {
    var index = digits.count - 1
    var number = 0
    var outDigits = [Int]()

    for digit in digits {
      number += digit * pow(base: baseIn, exponent: index)
      index -= 1
    }

    while number > 0 {
      outDigits.append(number % baseOut)
      number /= baseOut
    }
    
    return outDigits.reversed()
  }

  public static func
    outputDigits(inputBase: Int, inputDigits: [Int],
                 outputBase: Int) throws -> [Int] {
    if inputBase <= 1 {
      throw BaseError.invalidInputBase
    } else if outputBase <= 1 {
      throw BaseError.invalidOutputBase
    } else if inputDigits.contains(where: { $0 >= inputBase }) {
      throw BaseError.invalidPositiveDigit
    } else if inputDigits.contains(where: { $0 < 0 }) {
      throw BaseError.negativeDigit
    }

    return digits(baseIn: inputBase, digits: inputDigits, baseOut: outputBase)
  }
}

File: python/all-your-base/all_your_base.py
def power(base, exponent):
    if (exponent == 0):
        return 1
    n = 1
    for i in range(1, exponent):
        n *= base
    
    return n

def rebase(input_base, digits, output_base):
    index = len(digits) - 1
    number = 0
    outDigits = []
    
    for digit in digits:
        number += digit * pow(input_base, index)
        index -= 1 

    while number > 0:
        outDigits.append(number % output_base)
        number /= output_base
    
    return outDigits.reverse()